src/ServiceControl.Audit/Infrastructure/RavenDB/MigrateKnownEndpoints.cs

```diff
@@ -0,0 +1,90 @@
+        {
+            var knownEndpointsIndex = await store.AsyncDatabaseCommands.GetIndexAsync("EndpointsIndex").ConfigureAwait(false);
+            if (knownEndpointsIndex == null)
+            {
+                // Index has already been deleted, no need to migrate
+                return;
+            }
+
+            var dbStatistics = await store.AsyncDatabaseCommands.GetStatisticsAsync().ConfigureAwait(false);
+            var indexStats = dbStatistics.Indexes.First(index => index.Name == knownEndpointsIndex.Name);
+            if (indexStats.Priority == IndexingPriority.Disabled)
+            {
+                // This should only happen the second time the migration is attempted.
+                // The index is disabled so the data should have been migrated. We can now delete the index.
+                await store.AsyncDatabaseCommands.DeleteIndexAsync(knownEndpointsIndex.Name).ConfigureAwait(false);
+                return;
+            }
+
+            int previouslyDone = 0;
+            do
+            {
+                using (var session = store.OpenAsyncSession())
+                {
+                    var endpointsFromIndex = await session.Query<dynamic>(knownEndpointsIndex.Name, true)
+                        .Skip(previouslyDone)
+                        .Take(pageSize)
+                        .ToListAsync()
+                        .ConfigureAwait(false);
+
+                    if (endpointsFromIndex.Count == 0)
+                    {
+                        break;
+                    }
+                     
+                    previouslyDone += endpointsFromIndex.Count;
+
+                    var knownEndpoints = endpointsFromIndex.Select(endpoint => new KnownEndpoint
+                    {
+                        Id = KnownEndpoint.MakeDocumentId(endpoint.Name, Guid.Parse(endpoint.HostId)),
+                        Host = endpoint.Host,
+                        HostId = Guid.Parse(endpoint.HostId),
+                        Name = endpoint.Name,
+                        LastSeen = DateTime.UtcNow // Set the imported date to be now since we have no better guess
+                    });
+
+                    using (var bulkInsert = store.BulkInsert(options: new BulkInsertOptions
+                    {
+                        OverwriteExisting = true
+                    }))
+                    {
+                        foreach (var endpoint in knownEndpoints)
+                        {
+                            bulkInsert.Store(endpoint);
+                        }
+
+                        await bulkInsert.DisposeAsync().ConfigureAwait(false);
+                    }
+                }
+            } while (true);
+
+            // Disable the index so it can be safely deleted in the next migration run
+            await store.AsyncDatabaseCommands.SetIndexPriorityAsync(knownEndpointsIndex.Name, IndexingPriority.Disabled).ConfigureAwait(false);
```

Comment:
![image](https://user-images.githubusercontent.com/174258/90056853-d524d500-dcdf-11ea-95c6-b901d55ea4a1.png)

Say you can query a disabled index. If that is also true for 3.5 we might even consider switching the order? What do you think?

Reply:
Then you could disable the index, fail, run again, and the index would get deleted with no conversion.

Reply:
Good point, so obvious. Wonder why I haven't thought about that.